datanode/src/slippy_util.py

```diff
@@ -0,0 +1,322 @@
+
+
+def validate_slippy(z, x, y, raise_error=False):
+  """Validates slippy tile ranges.
+
+  https://en.wikipedia.org/wiki/Tiled_web_map
+  https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames
+
+  Args:
+    z: zoom level in slippy tile format
+    x: x tile number in slippy tile format
+    y: y tile number in slippy tile format
+  Returns:
+    true if valid, false if not
+  """
+  try:
+    z = int(z)
+    x = int(x)
+    y = int(y)
+    if not 0 <= z <= 20:
+      raise ValueError('Zoom must be an integer from 0-20')
+    if not 0 <= x < 2**z:
+      raise ValueError('x must be an integer from 0-%d' % (2**z - 1))
+    if not 0 <= y < 2**z:
+      raise ValueError('y must be an integer from 0-%d' % (2**z - 1))
+    return True
+  except (ValueError, TypeError) as e:
+    log.error('Invalid slippy format for tiles %sz, %s, %s!',
+              str(z), str(x), str(y))
+    log.error('Invalid slippy format error: %s', e)
+    if raise_error:
+      raise e
+    else:
+      return False
+
+
+def convert_csv_to_coordinates(csv):
+  """Converts and validates string of CSV coords into array of lat/lon coords."""
+  if not csv:
+    raise TypeError('CSV of coordinates must be a string in lat,lon[,lat,lon]...format')
+  result = []
+  coords = csv.split(',')
+  if len(coords) % 2 != 0:
+    raise ValueError('CSV of coordinates must in lat,lon pairs')
+  log.debug('Split coordinates to %s and passed early validation...', coords)
+  for a, b in _pairwise(coords):
+    lat = float(a)
+    lon = float(b)
+    if lat >= 90 or lat <= -90 or lon >= 180 or lon <= -180:
+      raise ValueError('CSV of coordinates must have valid lat,lon values')
+    result.append((lat, lon))
+  return result
+
+
+def convert_tile_to_polygon(zoom, xtile, ytile):
+  """Conversion from tile to lat,lon polygon at specific zoom.
+
+  Args:
+      zoom: level of zoom in slippy terms (0-20)
+      xtile: x tile in slippy terms
+      ytile: y tile in slippy terms
+  Raises:
+      TypeError if parameters are not ints
+      ValueError if parameters are not in respective bounds
+  Returns:
+      list of points that make up the tile
+  """
+  validate_slippy(zoom, xtile, ytile, True)
+  n = 2.0 ** zoom
+  wlon = xtile / n * 360.0 - 180.0
+  nlat = math.degrees(math.atan(math.sinh(math.pi * (1 - 2 * ytile / n))))
+  elon = (xtile + 1) / n * 360.0 - 180.0
+  slat = math.degrees(math.atan(math.sinh(math.pi * (1 - 2 * (ytile + 1) / n))))
+  return [(nlat, wlon), (nlat, elon), (slat, elon), (slat, wlon), (nlat, wlon)]
+
+
+def convert_point_to_tile(zoom, latitude, longitude):
+  """Conversion from lat/lon to tile at specific zoom.
+
+  Args:
+      zoom: level of zoom in slippy terms (0-20)
+      latitude: decimal degrees latitude for the tile
+      longitude: decimal degrees latitude for the tile
+  Raises:
+      TypeError if zoom is not integer or lat/on is not floats
+      ValueError if parameters are not in respective bounds
+  """
```

Comment:
nit: would be nice to note return values/format

Reply:
Done.